README
==== NAME ====

LCFG::Build::PkgSpec - Object-oriented interface to LCFG build metadata

==== DESCRIPTION ====

This class provides an object-oriented interface to the LCFG build
tools metadata file. All simple fields are available through attribute
accessors. Specific methods are also provided for querying and
modifying the more complex data types (e.g. lists and hashes).

This class has methods for carrying out specific procedures related to
tagging releases with the LCFG build tools. It also has methods for
handling the old format LCFG build configuration files.

More information on the LCFG build tools is available from the website
http://www.lcfg.org/doc/buildtools/


==== DEPENDENCIES ====

This module needs perl 5.8.1 or newer (that's for Moose nothing in
this code needs it).

To build this module you will need the following:

        Module::Build

To use this module you will need the following:

        DateTime
        Data::Structure::Util >= 0.12
        Moose >= 0.57
        MooseX::AttributeHelpers >= 0.13
        YAML::Syck >= 0.98

==== INSTALLATION ====

Make sure you have the dependencies installed first! (see DEPENDENCIES above)

If you don't know how to install these, try using the CPAN module, an easy
way of auto-installing modules from the Comprehensive Perl Archive Network,
where the above modules reside. Do "perldoc perlmodinstall" or "perldoc
CPAN" for more information.

To install this module type the following:

   perl Build.PL
   ./Build
   ./Build test
   ./Build install

Or, if you're on a platform (like DOS or Windows) that doesn't like the "./"
notation, you can do this:

   perl Build.PL
   perl Build
   perl Build test
   perl Build install

In order to install somewhere other than the default, such as in a directory
under your home directory, like "/home/fred/perl" go

   perl Build.PL --install_base /home/fred/perl

as the first step instead.

This will install the files underneath /home/fred/perl.

You will also need to make sure that you alter the PERL5LIB variable
to find the modules, and the PATH variable to find the script.

Therefore you will need to change: your path, to include
/home/fred/perl/script (where the script will be)

        export PATH=/home/fred/perl/script:${PATH}

the PERL5LIB variable to add /home/fred/perl/lib

        export PERL5LIB=/home/fred/perl/lib:${PERL5LIB}

If you want to install in a temporary install directory (such as if you are
building a package) then instead of going

   perl Build install

go

   perl Build install destdir=/my/temp/dir

and it will be installed there, with a directory structure under
/my/temp/dir the same as it would be if it were installed plain. Note that
this is NOT the same as setting --install_base, because certain things are
done at build-time which use the install_base info.

See "perldoc perlrun" for more information on PERL5LIB, and see "perldoc
Module::Build" for more information on installation options.

==== PROBLEMS AND FEEDBACK ====

There are no known bugs in this application. Please report any
problems to bugs@lcfg.org, feedback and patches are also always very
welcome.

==== AUTHOR ====

Stephen Quinney <squinney@inf.ed.ac.uk>

==== COPYRIGHT AND LICENCE ====

Copyright (C) 2008 University of Edinburgh

This program is free software; you can redistribute it and/or modify it
under the terms of the GPL, version 2 or later.



Changes
2013-01-22  SVN: new release

	* Release: 0.0.36

2013-01-22 14:19  squinney@INF.ED.AC.UK

	* bin/lcfg-pkgcfg.in: Reworked the get code some more so that
	  lookups for unsupported attributes give a different error code.
	  Also tidied the code slightly to appease perlcritic

2013-01-22 12:03  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-PkgSpec release: 0.0.35

2013-01-22 12:03  squinney@INF.ED.AC.UK

	* bin/lcfg-pkgcfg.in, lcfg.yml: Added support for accessing values
	  stored in hashes

2012-04-30 13:49  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-PkgSpec release: 0.0.34

2012-04-30 13:49  squinney@INF.ED.AC.UK

	* lcfg.yml, t/expected.yml: Updated test for new attribute

2012-04-30 13:47  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-PkgSpec release: 0.0.33

2012-04-30 13:47  squinney@INF.ED.AC.UK

	* bin/lcfg-pkgcfg.in, lib/LCFG/Build/PkgSpec.pm.in: Added new
	  orgident attribute and pkgident method. These are used for MacOSX
	  packaging

2011-02-14 17:05  squinney@INF.ED.AC.UK

	* README: updated README

2011-02-14 16:57  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-PkgSpec release: 0.0.32

2011-02-14 16:57  squinney@INF.ED.AC.UK

	* LCFG-Build-PkgSpec.spec: Fixed build-requires for new
	  Email::Address and Email::Valid dependencies

2011-02-14 16:52  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-PkgSpec release: 0.0.31

2011-02-14 16:45  squinney@INF.ED.AC.UK

	* Build.PL.in, META.yml.in, Makefile.PL,
	  lib/LCFG/Build/PkgSpec.pm.in, t/06_pkgspec_cfg.t: Switched to
	  using Native Traits so we can drop the dependency on the
	  deprecated AttributeHelpers module. Switched the email address
	  list handling for the authors field to Email::Address &
	  Email::Valid (new deps)

2011-02-14 16:41  squinney@INF.ED.AC.UK

	* Build.PL.in, MANIFEST, META.yml.in, Makefile.PL: Updated
	  dependency list, meta-data and build scripts

2011-02-14 15:55  squinney@INF.ED.AC.UK

	* t/10_pkgspec_date.t: fixed date spec tests so they always work

2010-11-19 10:43  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-PkgSpec release: 0.0.30

2010-11-19 10:43  squinney@INF.ED.AC.UK

	* lcfg.yml, lib/LCFG/Build/PkgSpec.pm.in, t/10_pkgspec_date.t:
	  Fixed the date formatting so that it is DD/MM/YY and not the
	  American style of MM/DD/YY. Also added tests so this does not
	  happen again

2009-03-13 15:14  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-PkgSpec release: 0.0.29

2009-03-13 14:53  squinney@INF.ED.AC.UK

	* Build.PL.in, LCFG-Build-PkgSpec.spec, META.yml.in, Makefile.PL,
	  README, lib/LCFG/Build/PkgSpec.pm.in: Switched from Date::Format
	  to DateTime

2009-03-11 13:25  squinney@INF.ED.AC.UK

	* bin/lcfg-cfg2meta.in, bin/lcfg-pkgcfg.in,
	  lib/LCFG/Build/PkgSpec.pm.in: Set svn:keywords on the
	  LCFG::Build::PkgSpec Perl modules and scripts

2009-03-09 16:43  squinney

	* lcfg.yml: Removed hardwired version-control type from lcfg.yml to
	  allow future transfer to subversion

2009-03-09 12:44  squinney

	* Changes, lcfg.yml: Release: 0.0.28

2009-03-09 12:44  squinney

	* lcfg.yml, t/06_pkgspec_cfg.t: Updated tests in 06_pkgspec_cfg.t
	  to reflect the fact that we no longer hardwire the
	  version-control system type

2009-03-09 12:42  squinney

	* Changes, lcfg.yml: Release: 0.0.27

2009-03-09 12:42  squinney

	* lib/LCFG/Build/PkgSpec.pm.in: lib/LCFG/Build/PkgSpec.pm.in
	  (new_from_cfgmk): No longer hardwire the version-control system
	  type to be CVS when converting config.mk files

2008-11-12 14:26  squinney

	* Changes, lcfg.yml: Release: 0.0.26

2008-11-12 14:26  squinney

	* t/09_clone.t: Switched to using is_deeply() in 09_clone.t test.
	  The eq_or_diff() method in Test::Differences does not handle
	  comparing an integer with a string (e.g. 1 is considered to not
	  equal '1'. See http://rt.cpan.org/Public/Bug/Display.html?id=3029

2008-11-12 14:24  squinney

	* lib/LCFG/Build/PkgSpec.pm.in: The schema is now explicitly an
	  integer.

2008-10-29 14:24  squinney

	* Changes, lcfg.yml: Release: 0.0.25

2008-10-29 14:24  squinney

	* lib/LCFG/Build/PkgSpec.pm.in: Fixed a small problem with parsing
	  the NAME field in a config.mk file

2008-09-12 14:28  squinney

	* Changes, lcfg.yml: Release: 0.0.24

2008-09-12 14:28  squinney

	* bin/lcfg-cfg2meta.in, bin/lcfg-pkgcfg.in: Switched file path
	  handling in lcfg-cfg2meta and lcfg-pkgcfg to using File::Spec

2008-09-10 13:59  squinney

	* bin/lcfg-pkgcfg.in, lib/LCFG/Build/PkgSpec.pm.in: Small
	  documentation tweaks to get links to other modules automatically
	  generated

2008-09-09 14:30  squinney

	* Changes, lcfg.yml: Release: 0.0.23

2008-09-09 14:29  squinney

	* Build.PL.in, LCFG-Build-PkgSpec.spec, META.yml, META.yml.in,
	  Makefile.PL, README, bin/lcfg-cfg2meta.in, bin/lcfg-pkgcfg.in,
	  lcfg.yml, lib/LCFG/Build/PkgSpec.pm.in, t/04_pkgspec_badattr.t:
	  Support fields in config.mk which use the ling-continuation
	  backslash. Improved handling of attributes which can be either
	  strings or array-refs so that they get properly transformed into
	  array-refs. Removed the override of the new() method as it is no
	  longer required. Made the LCFG::Build::PkgSpec class immutable to
	  get a speed gain. Improved the README and other docs. Updated
	  various Moose dependencies to ensure the code works correctly

2008-07-23 10:56  squinney

	* Changes, lcfg.yml: Release: 0.0.22

2008-07-23 10:56  squinney

	* bin/lcfg-cfg2meta.in: If not specified search the current working
	  directory for the config.mk when using lcfg-cfg2meta

2008-07-01 14:37  squinney

	* Changes, lcfg.yml: Release: 0.0.21

2008-07-01 14:36  squinney

	* t/06_pkgspec_cfg.t: Updated the tests which go with the config.mk
	  parser

2008-07-01 14:32  squinney

	* Changes, lcfg.yml: Release: 0.0.20

2008-07-01 14:32  squinney

	* lib/LCFG/Build/PkgSpec.pm.in: When parsing a config.mk make the
	  default license GPLv2 to be redhat compatible, also removed the

Build.PL
use Module::Build;

my $build = Module::Build->new(
    module_name   => 'LCFG::Build::PkgSpec',
    dist_abstract => 'Object-oriented interface to LCFG build metadata',
    dist_author   => 'Stephen Quinney <squinney@inf.ed.ac.uk>',
    dist_version  => '0.0.36',
    license => 'gpl',
    requires => {
        'perl'                      => '5.6.1',
        'Data::Structure::Util'     => '0.12',
        'DateTime'                  => '0',
        'Email::Address'            => '0',
        'Email::Valid'              => '0',
        'Moose'                     => '0.98',
        'Pod::Usage'                => '0',
        'YAML::Syck'                => '0.98',
    },
    build_requires => {
        'Test::More'        => 0,
        'Test::Differences' => 0,
        'Test::Exception'   => 0,
    },
    add_to_cleanup => [ 'LCFG-Build-PkgSpec-*', 'MANIFEST.bak', 'Changes.bak' ],
    create_makefile_pl => 'traditional',
);
$build->create_build_script;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.03
use ExtUtils::MakeMaker;
WriteMakefile
(
          'PL_FILES' => {},
          'INSTALLDIRS' => 'site',
          'NAME' => 'LCFG::Build::PkgSpec',
          'EXE_FILES' => [
                           'bin/lcfg-cfg2meta',
                           'bin/lcfg-pkgcfg'
                         ],
          'VERSION_FROM' => 'lib/LCFG/Build/PkgSpec.pm',
          'PREREQ_PM' => {
                           'Test::Exception' => 0,
                           'Pod::Usage' => '0',
                           'YAML::Syck' => '0.98',
                           'DateTime' => 0,
                           'Test::More' => 0,
                           'Test::Differences' => 0,
                           'MooseX::AttributeHelpers' => '0.13',
                           'Data::Structure::Util' => '0.12',
                           'Moose' => '0.57'
                         }
        )
;
